File: ios/TacticalMaps/Waypoints/Waypoint.swift
import Foundation
import CoreLocation
import SwiftUI

/// A user-placed point of interest. Stored in WGS84 as lat/lon doubles (matches
/// the GeoJSON export schema and the Android model). The computed `coordinate`
/// adapts to CoreLocation/MapKit APIs.
struct Waypoint: Identifiable, Codable, Hashable {
    let id: UUID
    var name: String
    var notes: String?
    var latitude: Double
    var longitude: Double
    var elevation: Double?      // metres above sea level (optional)
    var kind: WaypointKind
    /// Symbol rotation in degrees (0–360, clockwise). Only meaningful for
    /// tactical control measures whose orientation conveys direction
    /// (axis of advance, ambush, attack-by-fire, etc.). Ignored otherwise.
    var rotation: Double
    /// Horizontal multiplier on the symbol's default render size.
    /// 1.0 = canonical (~64pt on screen). Range surfaced in the UI is
    /// 0.1–20×. Independent of `scaleY` so the user can stretch a
    /// task graphic wider/thinner. Only applied to tactical control
    /// measures.
    var scaleX: Double
    /// Vertical multiplier. See `scaleX`.
    var scaleY: Double
    /// Tint applied to a tactical task graphic (control measure). The
    /// black line-art glyph is recoloured to this. Black is the default;
    /// the others follow the APP-6 affiliation palette. Ignored for
    /// military units and generic markers.
    var taskColor: TaskColor
    /// Which map layer this waypoint belongs to. Shared layer model with
    /// `DrawingShape` so toggling a layer hides both drawings and
    /// waypoints on it. Backward-compat: pre-layer saves get the
    /// `DrawingLayer.legacyFallbackID` so they land in the default
    /// "Friendly" layer.
    var layerID: UUID
    var createdAt: Date

    init(id: UUID = UUID(),
         name: String,
         notes: String? = nil,
         latitude: Double,
         longitude: Double,
         elevation: Double? = nil,
         kind: WaypointKind = .generic,
         rotation: Double = 0,
         scaleX: Double = 1.0,
         scaleY: Double = 1.0,
         taskColor: TaskColor = .black,
         layerID: UUID = DrawingLayer.legacyFallbackID,
         createdAt: Date = .now) {
        self.id = id
        self.name = name
        self.notes = notes
        self.latitude = latitude
        self.longitude = longitude
        self.elevation = elevation
        self.kind = kind
        self.rotation = rotation
        self.scaleX = scaleX
        self.scaleY = scaleY
        self.taskColor = taskColor
        self.layerID = layerID
        self.createdAt = createdAt
    }

    /// Convenience for callers working with CoreLocation/MapKit.
    init(id: UUID = UUID(),
         name: String,
         notes: String? = nil,
         coordinate: CLLocationCoordinate2D,
         elevation: Double? = nil,
         kind: WaypointKind = .generic,
         rotation: Double = 0,
         scaleX: Double = 1.0,
         scaleY: Double = 1.0,
         taskColor: TaskColor = .black,
         layerID: UUID = DrawingLayer.legacyFallbackID,
         createdAt: Date = .now) {
        self.init(id: id, name: name, notes: notes,
                  latitude: coordinate.latitude, longitude: coordinate.longitude,
                  elevation: elevation, kind: kind, rotation: rotation,
                  scaleX: scaleX, scaleY: scaleY, taskColor: taskColor,
                  layerID: layerID, createdAt: createdAt)
    }

    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }

    var subtitle: String? {
        elevation.map { String(format: "%.0f m", $0) }
    }

    /// Compact identity of `kind` used by MapContainerView's refresh
    /// fingerprint. Two waypoints with the same fingerprint render to
    /// the same annotation image, so the map can skip rebuilding.
    var kindFingerprint: String { kind.fingerprint }

    // MARK: Codable (custom to allow back-compat with files that pre-date `rotation`)

    private enum CodingKeys: String, CodingKey {
        case id, name, notes, latitude, longitude, elevation, kind,
             rotation, scale, scaleX, scaleY, taskColor, layerID, createdAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try c.decode(UUID.self, forKey: .id)
        self.name = try c.decode(String.self, forKey: .name)
        self.notes = try c.decodeIfPresent(String.self, forKey: .notes)
        self.latitude = try c.decode(Double.self, forKey: .latitude)
        self.longitude = try c.decode(Double.self, forKey: .longitude)
        self.elevation = try c.decodeIfPresent(Double.self, forKey: .elevation)
        self.kind = try c.decode(WaypointKind.self, forKey: .kind)
        self.rotation = try c.decodeIfPresent(Double.self, forKey: .rotation) ?? 0
        // Migration: older saves had a single `scale` field. If
        // scaleX/scaleY aren't present, populate both from `scale`
        // (or default 1.0).
        let legacyScale = try c.decodeIfPresent(Double.self, forKey: .scale)
        self.scaleX = try c.decodeIfPresent(Double.self, forKey: .scaleX)
            ?? legacyScale ?? 1.0
        self.scaleY = try c.decodeIfPresent(Double.self, forKey: .scaleY)
            ?? legacyScale ?? 1.0
        self.taskColor = try c.decodeIfPresent(TaskColor.self, forKey: .taskColor) ?? .black
        self.layerID = try c.decodeIfPresent(UUID.self, forKey: .layerID)
            ?? DrawingLayer.legacyFallbackID
        self.createdAt = try c.decode(Date.self, forKey: .createdAt)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encode(name, forKey: .name)
        try c.encodeIfPresent(notes, forKey: .notes)
        try c.encode(latitude, forKey: .latitude)
        try c.encode(longitude, forKey: .longitude)
        try c.encodeIfPresent(elevation, forKey: .elevation)
        try c.encode(kind, forKey: .kind)
        try c.encode(rotation, forKey: .rotation)
        try c.encode(scaleX, forKey: .scaleX)
        try c.encode(scaleY, forKey: .scaleY)
        try c.encode(taskColor, forKey: .taskColor)
        try c.encode(layerID, forKey: .layerID)
        try c.encode(createdAt, forKey: .createdAt)
    }
}

/// What a waypoint represents. Three top-level cases:
///   - `.generic`            : plain field marker
///   - `.military(spec)`     : APP-6C unit symbol (affiliation × echelon × function)
///   - `.controlMeasure(…)`  : tactical point-symbol control measure (FUP, RV, LZ, etc.)
enum WaypointKind: Hashable, Codable {
    case generic
    case military(MilitarySymbolSpec)
    case controlMeasure(TacticalControlMeasure)

    // MARK: Tagged Codable

    private enum CodingKeys: String, CodingKey { case type, spec, control }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let type = try c.decode(String.self, forKey: .type)
        switch type {
        case "generic":
            self = .generic
        case "military":
            self = .military(try c.decode(MilitarySymbolSpec.self, forKey: .spec))
        case "controlMeasure":
            self = .controlMeasure(try c.decode(TacticalControlMeasure.self, forKey: .control))
        default:
            self = .generic   // safe fallback for old/unknown persisted data
        }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .generic:
            try c.encode("generic", forKey: .type)
        case .military(let spec):
            try c.encode("military", forKey: .type)
            try c.encode(spec, forKey: .spec)
        case .controlMeasure(let m):
            try c.encode("controlMeasure", forKey: .type)
            try c.encode(m, forKey: .control)
        }
    }

    // MARK: Display

    /// Short, human-friendly summary.
    var displayName: String {
        switch self {
        case .generic:                return "Waypoint"
        case .military(let spec):
            // e.g. "Friendly Infantry Platoon"
            let prefix = spec.affiliation.displayName
            let role   = spec.function == .unspecified ? "" : spec.function.displayName + " "
            return "\(prefix) \(role)\(spec.echelon.displayName)"
        case .controlMeasure(let m):  return m.displayName
        }
    }

    /// Two-line category label used in the edit sheet.
    var categoryDisplayName: String {
        switch self {
        case .generic:         return "Field Marker"
        case .military:        return "Military Unit (APP-6C)"
        case .controlMeasure:  return "Tactical Control Measure"
        }
    }

    // MARK: Symbol accessors

    /// Non-nil for military kinds — used by the map and picker icon view.
    var militarySpec: MilitarySymbolSpec? {
        if case .military(let s) = self { return s }
        return nil
    }

    /// Tactical control measure (if any).
    var controlMeasure: TacticalControlMeasure? {
        if case .controlMeasure(let m) = self { return m }
        return nil
    }

    /// SF Symbol fallback for kinds without a custom drawing
    /// (generic + tactical control measures).
    var sfSymbol: String {
        switch self {
        case .generic:                  return "mappin"
        case .military:                 return "shield.fill"   // unused once militarySpec is wired
        case .controlMeasure:           return "flag.fill"
        }
    }

    /// Tint used when the kind falls back to an SF Symbol pin.
    var tint: Color {
        switch self {
        case .generic:        return .yellow
        case .military:       return .blue
        case .controlMeasure: return .black
        }
    }

    /// Compact representation used by the map's refresh fingerprint —
    /// distinguishes kinds that render to different images so we can
    /// skip rebuilding when nothing visible changed.
    var fingerprint: String {
        switch self {
        case .generic:
            return "g"
        case .military(let s):
            return "m|\(s.affiliation.rawValue)|\(s.echelon.rawValue)|\(s.function.rawValue)|\(s.isHeadquarters)"
        case .controlMeasure(let m):
            return "c|\(m.rawValue)"
        }
    }
}

// MARK: - Task graphic colour

/// Colour applied to a tactical task graphic (control measure). The
/// bundled glyphs are pure-black line art on transparent; the renderer
/// template-tints them to this colour. Black is the default; the other
/// four follow the APP-6 affiliation palette (saturated for legibility
/// on both satellite imagery and imported PDFs — the affiliation frame
/// fills are pastel and too pale for line art). Kept in sync with
/// Android's `TaskColor`.
enum TaskColor: String, Codable, Hashable, CaseIterable {
    case black, blue, red, green, yellow

    /// Tint for the glyph.
    var color: Color {
        switch self {
        case .black:  return .black
        case .blue:   return Color(red: 0x0E / 255, green: 0x5F / 255, blue: 0xD8 / 255)  // #0E5FD8
        case .red:    return Color(red: 0xD8 / 255, green: 0x28 / 255, blue: 0x1F / 255)  // #D8281F
        case .green:  return Color(red: 0x1E / 255, green: 0x8A / 255, blue: 0x34 / 255)  // #1E8A34
        case .yellow: return Color(red: 0xE2 / 255, green: 0xA4 / 255, blue: 0x00 / 255)  // #E2A400
        }
    }

    /// Picker label — colour plus its APP-6 affiliation meaning.
    var label: String {
        switch self {
        case .black:  return "Black"
        case .blue:   return "Blue (Friendly)"
        case .red:    return "Red (Hostile)"
        case .green:  return "Green (Neutral)"
        case .yellow: return "Yellow (Unknown)"
        }
    }
}

// MARK: - Tactical control measures (point-symbol subset of APP-6C)

/// Tactical mission tasks and control-measure point symbols.
/// Each case maps to a bundled PNG/SVG asset under
/// `Assets.xcassets/AppSymbols/<rawValue>`. Symbols are cropped from
/// NATO MPSOTC Table E-I (Tactical graphics), except `assemblyArea`
/// and `formUpPoint` which are custom SVGs.
enum TacticalControlMeasure: String, Codable, Hashable, CaseIterable {
    // ---- Mission tasks (cropped from PDF spec) ----
    case block, breach, bypass, canalise, clear, contain
    case counterattack, counterattackByFire, delay, destroy
    case disrupt, fix, interdict, isolate, neutralise
    case occupy, penetrate, reliefInPlace, retain, secure
    case screen, `guard`, cover, seize, withdraw, withdrawUnderPressure
    // ---- Control-measure point symbols ----
    case landingZone, ccp, observationPostRecon
    case axisOfMainAttack, axisOfSupportingAttack
    case attackByFire, supportByFire, ambush
    case antipersonnelMinefield, turn
    // ---- Custom ----
    case assemblyArea, formUpPoint

    var displayName: String {
        switch self {
        case .block:                  return "Block"
        case .breach:                 return "Breach"
        case .bypass:                 return "Bypass"
        case .canalise:               return "Canalise"
        case .clear:                  return "Clear"
        case .contain:                return "Contain"
        case .counterattack:          return "Counter-Attack"
        case .counterattackByFire:    return "Counter-Attack by Fire"
        case .delay:                  return "Delay"
        case .destroy:                return "Destroy"
        case .disrupt:                return "Disrupt"
        case .fix:                    return "Fix"
        case .interdict:              return "Interdict"
        case .isolate:                return "Isolate"
        case .neutralise:             return "Neutralise"
        case .occupy:                 return "Occupy"
        case .penetrate:              return "Penetrate"
        case .reliefInPlace:          return "Relief in Place"
        case .retain:                 return "Retain"
        case .secure:                 return "Secure"
        case .screen:                 return "Screen"
        case .`guard`:                  return "Guard"
        case .cover:                  return "Cover"
        case .seize:                  return "Seize"
        case .withdraw:               return "Withdraw"
        case .withdrawUnderPressure:  return "Withdraw Under Pressure"
        case .landingZone:            return "Landing Zone"
        case .ccp:                    return "Casualty Collection Point"
        case .observationPostRecon:   return "Observation Post (Recon)"
        case .axisOfMainAttack:       return "Axis of Main Attack"
        case .axisOfSupportingAttack: return "Axis of Supporting Attack"
        case .attackByFire:           return "Attack by Fire"
        case .supportByFire:          return "Support by Fire"
        case .ambush:                 return "Ambush"
        case .antipersonnelMinefield: return "Anti-Personnel Minefield"
        case .turn:                   return "Turn"
        case .assemblyArea:           return "Assembly Area"
        case .formUpPoint:            return "Form-Up Point"
        }
    }

    /// Basename of the bundled image in `Assets.xcassets/AppSymbols/`.
    var assetName: String { rawValue }

    /// Picker order — alphabetised by `displayName` so the in-app list
    /// matches what a user would scan for. Mirrors Android's
    /// `TacticalControlMeasure.pickerEntries`.
    static let pickerEntries: [TacticalControlMeasure] =
        allCases.sorted { $0.displayName.localizedCaseInsensitiveCompare($1.displayName) == .orderedAscending }
}

File: android/app/src/main/java/com/tacticalmaps/waypoints/Waypoint.kt
package com.tacticalmaps.waypoints

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import java.util.UUID

/**
 * A user-placed point of interest, stored in WGS84 so it survives swapping basemaps
 * (OSM, satellite tiles, GeoPDF, calibrated PDF).
 */
@Serializable
data class Waypoint(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val notes: String? = null,
    val latitude: Double,
    val longitude: Double,
    @SerialName("elevation_m") val elevationMetres: Double? = null,
    val kind: WaypointKind = WaypointKind.Generic,
    val rotation: Double = 0.0,
    val scaleX: Double = 1.0,
    val scaleY: Double = 1.0,
    @SerialName("task_color") val taskColor: TaskColor = TaskColor.BLACK,
    @SerialName("layer_id") val layerId: String = DEFAULT_LAYER_ID,
    @SerialName("created_at_epoch_ms") val createdAt: Long = System.currentTimeMillis()
) {
    val elevationLabel: String? get() = elevationMetres?.let { "%.0f m".format(it) }

    companion object {
        const val DEFAULT_LAYER_ID = "default"
    }
}

@Serializable(with = WaypointKindSerializer::class)
sealed interface WaypointKind {
    val displayName: String
    val categoryDisplayName: String

    @Serializable
    data object Generic : WaypointKind {
        override val displayName: String = "Waypoint"
        override val categoryDisplayName: String = "Field Marker"
    }

    @Serializable
    data class Military(val spec: MilitarySymbolSpec = MilitarySymbolSpec()) : WaypointKind {
        override val displayName: String
            get() = buildString {
                append(spec.affiliation.displayName)
                append(' ')
                if (spec.function != SymbolFunction.UNSPECIFIED) {
                    append(spec.function.displayName)
                    append(' ')
                }
                append(spec.echelon.displayName)
            }
        override val categoryDisplayName: String = "Military Unit (APP-6C)"
    }

    @Serializable
    data class ControlMeasure(
        val measure: TacticalControlMeasure = TacticalControlMeasure.ASSEMBLY_AREA
    ) : WaypointKind {
        override val displayName: String get() = measure.displayName
        override val categoryDisplayName: String = "Tactical Task"
    }
}

object WaypointKindSerializer : KSerializer<WaypointKind> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("WaypointKind")

    override fun serialize(encoder: Encoder, value: WaypointKind) {
        val jsonEncoder = encoder as? JsonEncoder
            ?: error("WaypointKind can only be encoded as JSON")
        val json = jsonEncoder.json
        val element = buildJsonObject {
            when (value) {
                WaypointKind.Generic -> put("type", "generic")
                is WaypointKind.Military -> {
                    put("type", "military")
                    put("spec", json.encodeToJsonElement(value.spec))
                }
                is WaypointKind.ControlMeasure -> {
                    put("type", "controlMeasure")
                    put("control", json.encodeToJsonElement(value.measure))
                }
            }
        }
        jsonEncoder.encodeJsonElement(element)
    }

    override fun deserialize(decoder: Decoder): WaypointKind {
        val jsonDecoder = decoder as? JsonDecoder
            ?: error("WaypointKind can only be decoded from JSON")
        val element = jsonDecoder.decodeJsonElement()
        val json = jsonDecoder.json

        if (element is JsonPrimitive) {
            return legacyKind(element.contentOrNull)
        }

        val obj = element.jsonObject
        return when (obj["type"]?.jsonPrimitive?.contentOrNull) {
            "military" -> WaypointKind.Military(
                obj["spec"]?.let { json.decodeFromJsonElement<MilitarySymbolSpec>(it) }
                    ?: MilitarySymbolSpec()
            )
            "controlMeasure" -> WaypointKind.ControlMeasure(
                obj["control"]?.let { json.decodeFromJsonElement<TacticalControlMeasure>(it) }
                    ?: TacticalControlMeasure.ASSEMBLY_AREA
            )
            else -> WaypointKind.Generic
        }
    }

    private fun legacyKind(raw: String?): WaypointKind = when (raw) {
        "drop_zone" -> WaypointKind.ControlMeasure(TacticalControlMeasure.LANDING_ZONE)
        "observation" -> WaypointKind.ControlMeasure(TacticalControlMeasure.OBSERVATION_POST_RECON)
        else -> WaypointKind.Generic
    }
}

@Serializable
data class MilitarySymbolSpec(
    val affiliation: SymbolAffiliation = SymbolAffiliation.FRIEND,
    val echelon: SymbolEchelon = SymbolEchelon.PLATOON,
    val function: SymbolFunction = SymbolFunction.INFANTRY,
    val isHeadquarters: Boolean = false
)

/**
 * Colour applied to a tactical task graphic (control measure). The bundled
 * glyphs are pure-black line art on transparent; the icon factory tints them
 * to this colour via a SRC_IN PorterDuff filter. Black is the default; the
 * other four follow the APP-6 affiliation palette (saturated for legibility
 * on satellite imagery and imported PDFs — the affiliation frame fills are
 * pastel and too pale for line art). Kept in sync with iOS's `TaskColor`.
 */
@Serializable
enum class TaskColor(val displayName: String, val argb: Int) {
    @SerialName("black") BLACK("Black", 0xFF000000.toInt()),
    @SerialName("blue") BLUE("Blue (Friendly)", 0xFF0E5FD8.toInt()),
    @SerialName("red") RED("Red (Hostile)", 0xFFD8281F.toInt()),
    @SerialName("green") GREEN("Green (Neutral)", 0xFF1E8A34.toInt()),
    @SerialName("yellow") YELLOW("Yellow (Unknown)", 0xFFE2A400.toInt())
}

@Serializable
enum class SymbolAffiliation(val displayName: String, val fillColor: Int) {
    @SerialName("friend") FRIEND("Friendly", 0xFF80E0FF.toInt()),
    @SerialName("hostile") HOSTILE("Hostile", 0xFFFF8080.toInt()),
    @SerialName("neutral") NEUTRAL("Neutral", 0xFFAAFFAA.toInt()),
    @SerialName("unknown") UNKNOWN("Unknown", 0xFFFFFF80.toInt())
}

@Serializable
enum class SymbolEchelon(val displayName: String, val glyph: String) {
    @SerialName("team") TEAM("Team / Crew", "Ø"),
    @SerialName("section") SECTION("Section", "●"),
    @SerialName("platoon") PLATOON("Platoon", "●●●"),
    @SerialName("company") COMPANY("Company", "I"),
    @SerialName("battalionRegiment") BATTALION_REGIMENT("Battalion / Regiment", "II"),
    @SerialName("brigade") BRIGADE("Brigade", "X"),
    @SerialName("division") DIVISION("Division", "XX")
}

@Serializable
enum class SymbolFunction(val assetName: String, val displayName: String) {
    @SerialName("airDefence") AIR_DEFENCE("airDefence", "Air Defence"),
    @SerialName("ammunition") AMMUNITION("ammunition", "Ammunition"),
    @SerialName("antiTank") ANTI_TANK("antiTank", "Anti-Tank"),
    @SerialName("armour") ARMOUR("armour", "Armour"),
    @SerialName("artillery") ARTILLERY("artillery", "Artillery"),
    @SerialName("aviationFixed") AVIATION_FIXED("aviationFixed", "Aviation (Fixed-Wing)"),
    @SerialName("aviation") AVIATION("aviation", "Aviation (Rotary)"),
    @SerialName("bridging") BRIDGING("bridging", "Bridging"),
    @SerialName("cavalry") CAVALRY("cavalry", "Cavalry"),
    @SerialName("cbrn") CBRN("cbrn", "CBRN Defence"),
    @SerialName("css") CSS("css", "Combat Service Support"),
    @SerialName("electronicWarfare") ELECTRONIC_WARFARE("electronicWarfare", "Electronic Warfare"),
    @SerialName("engineer") ENGINEER("engineer", "Engineer"),
    @SerialName("eod") EOD("eod", "Explosive Ordnance Disposal"),
    @SerialName("infantry") INFANTRY("infantry", "Infantry"),
    @SerialName("maintenance") MAINTENANCE("maintenance", "Maintenance"),
    @SerialName("mechInfantry") MECH_INFANTRY("mechInfantry", "Mechanised Infantry"),
    @SerialName("medical") MEDICAL("medical", "Medical"),
    @SerialName("militaryPolice") MILITARY_POLICE("militaryPolice", "Military Police"),
    @SerialName("mortar") MORTAR("mortar", "Mortar"),
    @SerialName("motorisedInfantry") MOTORISED_INFANTRY("motorisedInfantry", "Motorised Infantry"),
    @SerialName("radar") RADAR("radar", "Radar"),
    @SerialName("recce") RECCE("recce", "Reconnaissance"),
    @SerialName("signal") SIGNAL("signal", "Signals"),
    @SerialName("specialForces") SPECIAL_FORCES("specialForces", "Special Forces"),
    @SerialName("logistics") LOGISTICS("logistics", "Supply"),
    @SerialName("transportation") TRANSPORTATION("transportation", "Transportation"),
    @SerialName("uav") UAV("uav", "Unmanned Air Vehicle"),
    @SerialName("unspecified") UNSPECIFIED("unspecified", "No Branch");

    companion object {
        val pickerEntries: List<SymbolFunction>
            get() = entries
                .filterNot { it == UNSPECIFIED }
                .sortedBy { it.displayName } + UNSPECIFIED
    }
}

@Serializable
enum class TacticalControlMeasure(val assetName: String, val displayName: String) {
    @SerialName("block") BLOCK("block", "Block"),
    @SerialName("breach") BREACH("breach", "Breach"),
    @SerialName("bypass") BYPASS("bypass", "Bypass"),
    @SerialName("canalise") CANALISE("canalise", "Canalise"),
    @SerialName("clear") CLEAR("clear", "Clear"),
    @SerialName("contain") CONTAIN("contain", "Contain"),
    @SerialName("counterattack") COUNTERATTACK("counterattack", "Counter-Attack"),
    @SerialName("counterattackByFire") COUNTERATTACK_BY_FIRE("counterattackByFire", "Counter-Attack by Fire"),
    @SerialName("delay") DELAY("delay", "Delay"),
    @SerialName("destroy") DESTROY("destroy", "Destroy"),
    @SerialName("disrupt") DISRUPT("disrupt", "Disrupt"),
    @SerialName("fix") FIX("fix", "Fix"),
    @SerialName("interdict") INTERDICT("interdict", "Interdict"),
    @SerialName("isolate") ISOLATE("isolate", "Isolate"),
    @SerialName("neutralise") NEUTRALISE("neutralise", "Neutralise"),
    @SerialName("occupy") OCCUPY("occupy", "Occupy"),
    @SerialName("penetrate") PENETRATE("penetrate", "Penetrate"),
    @SerialName("reliefInPlace") RELIEF_IN_PLACE("reliefInPlace", "Relief in Place"),
    @SerialName("retain") RETAIN("retain", "Retain"),
    @SerialName("secure") SECURE("secure", "Secure"),
    @SerialName("screen") SCREEN("screen", "Screen"),
    @SerialName("guard") GUARD("guard", "Guard"),
    @SerialName("cover") COVER("cover", "Cover"),
    @SerialName("seize") SEIZE("seize", "Seize"),
    @SerialName("withdraw") WITHDRAW("withdraw", "Withdraw"),
    @SerialName("withdrawUnderPressure") WITHDRAW_UNDER_PRESSURE("withdrawUnderPressure", "Withdraw Under Pressure"),
    @SerialName("landingZone") LANDING_ZONE("landingZone", "Landing Zone"),
    @SerialName("ccp") CCP("ccp", "Casualty Collection Point"),
    @SerialName("observationPostRecon") OBSERVATION_POST_RECON("observationPostRecon", "Observation Post (Recon)"),
    @SerialName("axisOfMainAttack") AXIS_OF_MAIN_ATTACK("axisOfMainAttack", "Axis of Main Attack"),
    @SerialName("axisOfSupportingAttack") AXIS_OF_SUPPORTING_ATTACK("axisOfSupportingAttack", "Axis of Supporting Attack"),
    @SerialName("attackByFire") ATTACK_BY_FIRE("attackByFire", "Attack by Fire"),
    @SerialName("supportByFire") SUPPORT_BY_FIRE("supportByFire", "Support by Fire"),
    @SerialName("ambush") AMBUSH("ambush", "Ambush"),
    @SerialName("antipersonnelMinefield") ANTIPERSONNEL_MINEFIELD("antipersonnelMinefield", "Anti-Personnel Minefield"),
    @SerialName("turn") TURN("turn", "Turn"),
    @SerialName("assemblyArea") ASSEMBLY_AREA("assemblyArea", "Assembly Area"),
    @SerialName("formUpPoint") FORM_UP_POINT("formUpPoint", "Form-Up Point");

    companion object {
        val pickerEntries: List<TacticalControlMeasure>
            get() = entries.sortedBy { it.displayName }
    }
}
